### PR TITLE
[Storage] Fine grained account storage

### PR DESCRIPTION
**Pull request by sitalkedia**
## Motivation

This PR implements fine-grained storage for account state - instead of storing an account as a blob under one single key in the storage layer, we break it down into multiple resources, and each resource is stored as a blob with an access path as the key. The obvious advantage of this model is that, when accessing a resource from an account, we don't have access entire account - which can be expensive if the account is large. This will make accessing individual resources under an account cheaper in terms of gas cost. This model also allows Smart contract writes to organize the data to minimize the gas cost - they can choose to group items together under same resource which they expect to be used together - this will minimize the number of IO operations we need to perform during the txn execution.

Few more TODOs that I plan to address in subsequent PRs (as the scope of this change is already pretty big)

- [ ] Unify functions under `AccountState` and `AccountStateView and remove duplicates.
- [ ] Get rid of `AccountStateBlob` and related APIs. 

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

(You must have submitted a [signed CLA](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#contributor-license-agreement) that includes your GitHub handle prior to us accepting and landing your work. Write your answer here.)

## Test Plan

Existing Unit tests and added new tests.

### CODE REVIEW

File: storage/state-view/src/account_state_view.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        access_path_for_config, dpn_access_path_for_config, ConfigID, ConfigurationResource,
+        OnChainConfig, ValidatorSet,
+    },
+    state_store::state_key::StateKey,
+    validator_config::ValidatorConfigResource,
+};
+use move_core_types::{account_address::AccountAddress, move_resource::MoveResource};
+use serde::de::DeserializeOwned;
+
+pub struct AccountStateView<'a, F>
```
Comment (msmouse) [inline storage/state-view/src/account_state_view.rs:19]: This will do, right?
```
pub struct AccountStateView<'a, S> {
    address: AccountAddress,
    state_view: &'a S,
}
```

(`AccountAddress` is `copy`, so we don't really bother to use a ref, but it doesn't matter anyway I guess.)

File: config/management/genesis/src/verify.rs [hunk 2]
```diff
@@ -239,15 +238,17 @@ fn validator_config(
...
-    let blob = reader
-        .get_latest_state_value(StateKey::AccountAddressKey(
-            account_config::validator_set_address(),
-        ))
-        .map_err(|e| Error::UnexpectedError(format!("ValidatorSet Account issue {}", e)))?
-        .ok_or_else(|| Error::UnexpectedError("ValidatorSet Account not found".into()))?;
-    let account_state = AccountState::try_from(&blob)
-        .map_err(|e| Error::UnexpectedError(format!("Failed to parse blob: {}", e)))?;
-    let validator_set: ValidatorSet = account_state
+    let account_state_view = AccountStateView {
```
Comment (msmouse) [inline config/management/genesis/src/verify.rs:241]: This can be
```
let account_view = reader
    .lastest_state_view()
    .as_account_view(validator_set_address());
```
if you do it as my other comment suggests.

**Review by msmouse — APPROVED** [review]

**Comment by msmouse** [comment]
(sorry, I can't help approving your PRs. Then I had to put a red cross to reverse that 😂 )

**Review by msmouse — CHANGES_REQUESTED** [review]
(sorry, I can't help approving your PRs. Then I had to put a red cross to reverse that 😂 )

**Comment by sitalkedia** [comment]
This PR is ready for review. There are still few UTs failing - I am working on resolving them but that should not be a blocker for the review.

File: storage/storage-interface/src/lib.rs [hunk 3]
```diff
@@ -525,47 +536,62 @@ impl MoveStorage for &dyn DbReader {
...
+        let config_value_option = self
+            .get_state_value_with_proof_by_version(
+                &StateKey::AccessPath(AccessPath::new(
+                    aptos_root_address(),
+                    access_path_for_config(config_id).path,
+                )),
+                version,
+            )?
+            .0;
+        match config_value_option {
```
Comment (msmouse) [inline storage/storage-interface/src/lib.rs:558]: it's cumbersome to deal with the two `None` cases in  `Option<StateValue>`.. shall we create some sort of helper function for that??  For places where we don't care about the proofs, ideally the API can hide the details wrt "None as a placeholder for deleted value" situation..

File: testsuite/smoke-test/src/state_sync_v2.rs [hunk 4]
```diff
@@ -15,7 +16,7 @@ use std::{
     time::{Duration, Instant},
 };
 
-const MAX_CATCH_UP_SECS: u64 = 60; // The max time we'll wait for nodes to catch up
```
Comment (msmouse) [inline testsuite/smoke-test/src/state_sync_v2.rs:18]: unnecessary change

File: types/src/account_state_view.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    },
+    state_store::state_key::StateKey,
+    validator_config::ValidatorConfig,
+};
+use move_core_types::{
+    account_address, account_address::AccountAddress, move_resource::MoveResource,
+};
+use serde::de::DeserializeOwned;
+
+pub struct AccountStateView<'a, F>
```
Comment (msmouse) [inline types/src/account_state_view.rs:19]: Great idea to introduce this. 

(I'd just call it `AccountView` to distinguish from `StateView` and be shorter.)

Again, try to create a `trait AsAccountView` and impl it for all `S: StateView`, which will simplify the call sites.

File: vm-validator/src/unit_tests/vm_validator_test.rs [hunk 6]
```diff
@@ -193,6 +196,33 @@ fn test_validate_max_gas_units_below_min() {
         ret.status().unwrap(),
         StatusCode::MAX_GAS_UNITS_BELOW_MIN_TRANSACTION_GAS_UNITS
     );
+    let x = get_account_sequence_number(
+        vm_validator.vm_validator.db_reader,
+        AccountAddress::new([0u8; AccountAddress::LENGTH]),
+    )
+    .unwrap();
+    print!("x is {:?}", x);
```
Comment (msmouse) [inline vm-validator/src/unit_tests/vm_validator_test.rs:204]: remove

**Review by msmouse — COMMENTED** [review]
Gave it a first pass. We are definitely on the right path.

File: execution/executor/src/components/chunk_output.rs [hunk 7]
```diff
@@ -50,6 +63,17 @@ impl ChunkOutput {
...
+        // Warm up the state view cache by reading the configuration key, which is needed
+        // later when we parse the validator set and epoch.
+        state_view.get_state_value(&StateKey::AccessPath(AccessPath::new(
+            on_chain_config::config_address(),
+            ConfigurationResource::struct_tag().access_vector(),
+        )))?;
+        state_view.get_state_value(&StateKey::AccessPath(AccessPath::new(
+            on_chain_config::config_address(),
+            access_path_for_config(ValidatorSet::CONFIG_ID).path,
+        )))?;
```
[the remark this answers is not in the document]
Comment (zekun000) [reply execution/executor/src/components/chunk_output.rs:75]: I don't think we need to touch after the Stake.move, let me know if it doesn't work after you rebase. https://github.com/aptos-labs/aptos-core/blob/8f49b1190b9d4557c7cdbe1fc53b8108701585e0/aptos-move/framework/aptos-framework/sources/configs/Stake.move#L244

[the remark this answers is not in the document]
Comment (sitalkedia) [reply execution/executor/src/components/chunk_output.rs:75]: @zekun000 - Validator set doesn't need to be touched on the execution path but this is the state sync path where VM is not invoked IIUC. So we would need to touch this path right?

File: storage/storage-interface/src/lib.rs (hunk 3 above)
Comment (sitalkedia) [reply storage/storage-interface/src/lib.rs:558]: I like the idea of the API to simplify this. Can you clarify what you have in mind? Is this a helper function in StateValue, which takes `Option<StateValue>` and returns an `Option<Vec<u8>>`

File: execution/executor/src/components/chunk_output.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (zekun000) [reply execution/executor/src/components/chunk_output.rs:75]: it should be in the writeset as well for epoch change

File: storage/storage-interface/src/lib.rs [hunk 8]
```diff
@@ -525,47 +536,62 @@ impl MoveStorage for &dyn DbReader {
...
+                || {
+                    Err(format_err!(
+                        "no config {} found in aptos root account state",
+                        config_id
+                    ))
+                },
+                |bytes| Ok(bytes.to_vec()),
+            ),
+            _ => {
+                let dpn_config_value_option = self
```
Comment (zekun000) [inline storage/storage-interface/src/lib.rs:569]: I don't think we need the DPN branch any more

File: types/src/account_state_view.rs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    pub fn get_validator_config_resource(&self) -> anyhow::Result<Option<ValidatorConfig>> {
+        self.get_resource::<ValidatorConfig>()
+    }
+
+    fn get_on_chain_config<T: OnChainConfig>(&self) -> anyhow::Result<Option<T>> {
+        let state_key = self.get_state_key_for_path(access_path_for_config(T::CONFIG_ID).path);
+        match self.get_resource_impl(&state_key)? {
+            Some(config) => Ok(Some(config)),
+            _ => self.get_resource_impl(
+                &self.get_state_key_for_path(dpn_access_path_for_config(T::CONFIG_ID).path),
```
Comment (zekun000) [inline types/src/account_state_view.rs:63]: no need for this

File: execution/executor/src/components/chunk_output.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (sitalkedia) [reply execution/executor/src/components/chunk_output.rs:75]: @zekun000, @msmouse - Sounds good -Let me follow up with a change to extract the epoch and validator set from the writeset for epoch change. Are we okay with not doing that in this diff in order to not increase the scope of this change further?

File: execution/executor-benchmark/src/transaction_generator.rs [hunk 10]
```diff
@@ -331,7 +329,7 @@ impl TransactionGenerator {
     }
 
     /// Verifies the sequence numbers in storage match what we have locally.
-    pub fn verify_sequence_number(&self, db: &dyn DbReader) {
+    pub fn verify_sequence_number(&self, db: Arc<dyn DbReader>) {
```
Comment (zekun000) [inline execution/executor-benchmark/src/transaction_generator.rs:332]: any particular reason for this change?

File: storage/storage-interface/src/lib.rs [hunk 11]
```diff
@@ -642,6 +668,33 @@ pub trait DbWriter: Send + Sync {
...
+            return Ok(state_value.maybe_bytes.as_ref().cloned());
+        } else {
+            Ok(None)
+        }
+    })
+}
+
+pub fn get_state_value_resolver_for_latest_version(
+    db_reader: Arc<dyn DbReader>,
+) -> Box<dyn Fn(&StateKey) -> anyhow::Result<Option<Vec<u8>>>> {
```
Comment (zekun000) [inline storage/storage-interface/src/lib.rs:687]: why we need a closure for this? can't we use static function?

File: execution/executor/src/components/chunk_output.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (zekun000) [reply execution/executor/src/components/chunk_output.rs:75]: we don't need to extract anything from what I understand, remove the warm up part it should just work.

File: storage/storage-interface/src/lib.rs (hunk 11 above)
Comment (sitalkedia) [reply storage/storage-interface/src/lib.rs:687]: @zekun000 - `AccountStateView` needs to accept a closure to cater to various needs.  `AccountStateView` is currently used in many different scenarios as follows

1. StateCache (HashMap) to provide this view
2. DB reader to provide latest state view.
3. DB reader to provide state view for a version.
4. [TBD] AccountState to provide this view to remove redundant functions in AccountState.

Let me know if you have any better idea to structure this.

File: execution/executor-benchmark/src/transaction_generator.rs (hunk 10 above)
Comment (sitalkedia) [reply execution/executor-benchmark/src/transaction_generator.rs:332]: The call site already has Arc<dyn DbReader>, which also needs to be passed to the `get_state_value_resolver_for_latest_version` API

File: storage/storage-interface/src/lib.rs (hunk 3 above)
Comment (msmouse) [reply storage/storage-interface/src/lib.rs:558]: Yeah if you want to do it fancy: 

```
trait ToValueOption {
  pub fn to_value_opt()
}

impl ToValueOption for Option<StateValue> {
...
}
```

File: types/src/account_state_view.rs (hunk 5 above)
Comment (msmouse) [reply types/src/account_state_view.rs:18]: Introduce AccountViewTrait? 

```
trait AccountViewTrait {
    fn get();

    fn get_balance() {
        ... // blanket impl that calls self.get()
    }

    ...
}

impl AccountViewTrait for HashMap<.., ..>

impl AccountViewTrait for AccountState<.., ..>

impl AccountViewTrait for AccountView

struct AccountView<S: StateView> {
    address: AccountAddress,
    state_view: S
}
```

Comment (sitalkedia) [reply types/src/account_state_view.rs:18]: Hmm but that would mean a lot of code duplication as the actual implementations need to rewrite similar functions?

Comment (msmouse) [reply types/src/account_state_view.rs:18]: no the actual impl provide `get()` only.

Comment (sitalkedia) [reply types/src/account_state_view.rs:18]: @msmouse - Thanks for the suggestion. I actually had started with this approach but resorted to changing it to the current one because I faced some issues. While re-implementing your suggestion, I now recall the issue I ran into. So the issue is for DbReader, we need two implementation of AccountViewTrait one with version and other without version. AFAIK, rust won't allow something like

```
impl AccountView for (dyn DbReader, Version) {
```

It doesn't like trait implementation for arbitrary types. So one way to do that is create one more wrapper type with db_reader and version. The same issue exists for HashMap as well - we need to wrap the hashmap in another type to be able to define a trait implementation. I think this becomes less ergonomic as compared to current approach which is provides flexibility of taking arbitrary fn pointer. Let me know if you think defining the wrapper is a better way, we can do that.

File: execution/executor/src/components/chunk_output.rs [hunk 12]
```diff
@@ -34,6 +39,14 @@ impl ChunkOutput {
         transactions: Vec<Transaction>,
         state_view: VerifiedStateView,
     ) -> Result<Self> {
+        // Warm up the state view cache by reading the configuration key, which is needed
+        // later when we parse the validator set and epoch.
+        // This is a Hack currently and will follow up to fix this by reading the epoch from events
+        state_view.get_state_value(&StateKey::AccessPath(AccessPath::new(
+            on_chain_config::config_address(),
+            ConfigurationResource::struct_tag().access_vector(),
+        )))?;
```
[the remark this answers is not in the document]
Comment (sitalkedia) [reply execution/executor/src/components/chunk_output.rs:48]: That's great, after pulling in the latest Staking changes, we can get rid of this.

File: execution/executor/src/components/chunk_output.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (sitalkedia) [reply execution/executor/src/components/chunk_output.rs:75]: Thank you for doing this - simplify this a lot on storage side :D

File: types/src/account_state_view.rs (hunk 5 above)
Comment (msmouse) [reply types/src/account_state_view.rs:18]:
```
impl<T: DbReader> AccountViewTrait for (T, Version) {
}
```

^ you implement for arbitrary dbreaders like this.. but I do prefer the wrapper, (around `S: StateView`, not the dbreader.).


For things that are not a `StateView`, namely `AccountState` and the  `HashMap` we implement the trait directly for them.

**Comment by sitalkedia** [comment]
Hmm so forge TPS is ~1400 and I think with parallel execution enabled, we expect ~2400 TPS - thats around 40% regression. Is my understanding correct? cc - @msmouse @zekun000 ?

**Comment by zekun000** [comment]
looks like so, some graphs for comparison
http://o11y.aptosdev.com/grafana/d/consensus/consensus?orgId=1&from=1651034065314&to=1651036195758
http://o11y.aptosdev.com/grafana/d/system/system?orgId=1&from=1651034143162&to=1651036274426

**Comment by msmouse** [comment]
I'd rather ship this after the basic resources are combined. Not sure if we can make it in a week though.

**Comment by sitalkedia** [comment]
@msmouse - We can evaluate that option but my concern is the cost of maintaining this big change and rebasing on our very dynamic repo is very high.

File: storage/storage-interface/src/lib.rs (hunk 3 above)
Comment (sitalkedia) [reply storage/storage-interface/src/lib.rs]: Actually, nothing fancy is needed here.

```
  config_value_option
            .and_then(|x| x.maybe_bytes)
```
This will do.

File: types/src/account_state_view.rs (hunk 5 above)
Comment (sitalkedia) [reply types/src/account_state_view.rs]: >>For things that are not a StateView, namely AccountState and the HashMap we implement the trait directly for them.

We can't implement these traits directly on HashMap, I have defined a separate struct for doing that.

File: storage/storage-interface/src/lib.rs [hunk 13]
```diff
@@ -525,47 +536,62 @@ impl MoveStorage for &dyn DbReader {
...
-        let (state_value, _) = self.get_state_value_with_proof_by_version(
-            &StateKey::AccountAddressKey(access_path.address),
-            version,
-        )?;
-        let account_state =
-            AccountState::try_from(&state_value.ok_or_else(|| {
-                format_err!("missing blob in account state/account does not exist")
-            })?)?;
+        let (state_value, _) = self
+            .get_state_value_with_proof_by_version(&StateKey::AccessPath(access_path), version)?;
```
[the remark this answers is not in the document]
Comment (sitalkedia) [reply storage/storage-interface/src/lib.rs:540]: https://github.com/aptos-labs/aptos-core/issues/720 filed

File: storage/storage-interface/src/state_view.rs [hunk 14]
```diff
@@ -69,14 +68,6 @@ pub trait DbStateViewAtVersion {
 
 impl DbStateViewAtVersion for Arc<dyn DbReader> {
     fn state_view_at_version(&self, version: Option<Version>) -> Result<DbStateView> {
-        if let Some(some_version) = version {
```
Comment (sitalkedia) [inline storage/storage-interface/src/state_view.rs:72]: @msmouse - FYI, there is some inconsistency in this `self.get_latest_version()` is most likely stale in some cases and there were few test cases failing because of this check. In particular, `get_latest_transaction_info_option` returns a version which is higher than `get_latest_version`. So the `fetch_latest_synced_version`  function fails. This is something we might need to follow up - I will file a follow up issue for this.

Comment (msmouse) [reply storage/storage-interface/src/state_view.rs:72]: The problem is actually that the "synced version" can (reasonably) be ahead of the "latest version", which is a "committed version", which means there's a ledger info at or beyond that version. --- it's the correct thing to do to remove the check.

**Review by msmouse — APPROVED** [review]

File: storage/storage-interface/src/state_view.rs [hunk 15]
```diff
@@ -31,35 +29,36 @@ impl DbStateView {
...
+
+impl StateView for LatestDbStateView {
+    fn get_state_value(&self, state_key: &StateKey) -> Result<Option<Vec<u8>>> {
+        let value_opt = self.db.get_latest_state_value(state_key.clone())?;
+        if let Some(value) = value_opt {
+            Ok(value.maybe_bytes)
+        } else {
+            Ok(None)
+        }
+    }
```
Comment (msmouse) [inline storage/storage-interface/src/state_view.rs:58]: Please don't.. it's a crime to return inconsistent (meaning at different db versions) view of the db from something called a "StateView". I'd call it a bug. Why is this necessary?

**Review by msmouse — CHANGES_REQUESTED** [review]
sorry I haven't finished the final review. (I should write a browser extension to prevent myself from clicking "Approve" every time I want to make a comment.

File: storage/storage-interface/src/state_view.rs (hunk 15 above)
Comment (sitalkedia) [reply storage/storage-interface/src/state_view.rs]: Good point - this was needed for some tests. Found another way to fix that.

File: execution/executor/src/components/chunk_output.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (msmouse) [reply execution/executor/src/components/chunk_output.rs:75]: This as well.

File: execution/executor/src/mock_vm/mod.rs [hunk 16]
```diff
@@ -146,8 +145,14 @@ impl VMExecutor for MockVM {
...
-                    read_balance_from_storage(state_view, &balance_ap(validator_set_address()));
-                    read_balance_from_storage(state_view, &balance_ap(aptos_root_address()));
+                    read_state_value_from_storage(
+                        state_view,
+                        &access_path_for_config(ValidatorSet::CONFIG_ID),
+                    );
+                    read_state_value_from_storage(
+                        state_view,
+                        &AccessPath::new(config_address(), ConfigurationResource::resource_path()),
+                    );
```
Comment (msmouse) [inline execution/executor/src/mock_vm/mod.rs:155]: 🎉

File: storage/storage-interface/src/state_view.rs [hunk 17]
```diff
@@ -31,25 +33,23 @@ impl DbStateView {
...
 
+impl<'a> AsAccountWithStateView<'a> for DbStateView {
+    fn as_account_with_state_view(
+        &'a self,
+        account_address: &'a AccountAddress,
+    ) -> AccountWithStateView {
+        AccountWithStateView::new(account_address, self)
+    }
+}
+
```
Comment (msmouse) [inline storage/storage-interface/src/state_view.rs:52]: Doesn't this work?

```
impl<'a, S: 'a + StateView> AsAccountWithStateView for S {
    ...
}
```

**Review by msmouse — COMMENTED** [review]
I see aptossum is still on your stack.. I'm sorry but you need to rebase again.

File: types/src/account_state.rs [hunk 18]
```diff
@@ -4,21 +4,20 @@
...
-        VMPublishingOption, ValidatorSet, Version,
-    },
-    state_store::state_value::StateValue,
-    timestamp::TimestampResource,
+    on_chain_config::{access_path_for_config, OnChainConfig, ValidatorSet},
+    state_store::{state_key::StateKey, state_value::StateValue},
     validator_config::{ValidatorConfig, ValidatorOperatorConfigResource},
 };
-use anyhow::{format_err, Error, Result};
+use anyhow::{anyhow, format_err, Error, Result};
```
[the remark this answers is not in the document]
Comment (sitalkedia) [reply types/src/account_state.rs]: Removed

File: storage/storage-interface/src/state_view.rs (hunk 17 above)
Comment (sitalkedia) [reply storage/storage-interface/src/state_view.rs]: Haha, thank you for the fancier tip :)

**Comment by sitalkedia** [comment]
Deleted the warm-up hack as @zekun000's PR is already in.

File: storage/aptosdb/src/aptossum/mod.rs [hunk 19]
```diff
@@ -1,105 +0,0 @@
-// Copyright (c) Aptos
```
Comment (msmouse) [inline storage/aptosdb/src/aptossum/mod.rs:1]: weird... did you pull the latest?

**Review by msmouse — APPROVED** [review]

**Comment by msmouse** [comment]
Thanks. This approval is real. 🎃

File: storage/aptosdb/src/aptossum/mod.rs (hunk 19 above)
Comment (sitalkedia) [reply storage/aptosdb/src/aptossum/mod.rs:1]: Yes, this is the latest. I think https://github.com/aptos-labs/aptos-core/pull/714 missed to remove this file.

Comment (lightmark) [reply storage/aptosdb/src/aptossum/mod.rs:1]: Shit, my bad~
